File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/evaluation/AuroraReportEvaluator.java
package se.gustavkarlsson.aurora_notifier.android.evaluation;

import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import java8.util.stream.RefStreams;
import se.gustavkarlsson.aurora_notifier.android.models.AuroraFactors;
import se.gustavkarlsson.aurora_notifier.android.models.AuroraReport;

import static java8.util.stream.StreamSupport.stream;

public class AuroraReportEvaluator implements ChanceEvaluator<AuroraReport> {

	private final GeomagActivityEvaluator geomagActivityEvaluator;
	private final GeomagLocationEvaluator geomagLocationEvaluator;
	private final VisibilityEvaluator visibilityEvaluator;
	private final DarknessEvaluator darknessEvaluator;

	@Inject
	public AuroraReportEvaluator(GeomagActivityEvaluator geomagActivityEvaluator, GeomagLocationEvaluator geomagLocationEvaluator, VisibilityEvaluator visibilityEvaluator, DarknessEvaluator darknessEvaluator) {
		this.geomagActivityEvaluator = geomagActivityEvaluator;
		this.geomagLocationEvaluator = geomagLocationEvaluator;
		this.visibilityEvaluator = visibilityEvaluator;
		this.darknessEvaluator = darknessEvaluator;
	}

	@Override
	public Chance evaluate(AuroraReport auroraReport) {
		AuroraFactors factors = auroraReport.getFactors();
		Chance activityChance = geomagActivityEvaluator.evaluate(factors.getGeomagActivity());
		Chance locationChance = geomagLocationEvaluator.evaluate(factors.getGeomagLocation());
		Chance visibilityChance = visibilityEvaluator.evaluate(factors.getVisibility());
		Chance darknessChance = darknessEvaluator.evaluate(factors.getDarkness());

		List<Chance> chances = Arrays.asList(activityChance, locationChance, visibilityChance, darknessChance);

		if (stream(chances).anyMatch(c -> !c.isKnown())) {
			return Chance.unknown();
		}

		if (stream(chances).anyMatch(c -> !c.isPossible())) {
			return Chance.impossible();
		}

		Chance geomagActivityAndLocationChance = Chance.of(activityChance.getValue() * locationChance.getValue());

		return RefStreams.of(visibilityChance, darknessChance, geomagActivityAndLocationChance)
				.min(Chance::compareTo)
				.orElseThrow(() -> new IllegalStateException("No PresentableChance supplied"));
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/services_impl/evaluation/AuroraReportEvaluator.kt
package se.gustavkarlsson.skylight.android.services_impl.evaluation

import se.gustavkarlsson.skylight.android.entities.AuroraReport
import se.gustavkarlsson.skylight.android.entities.Chance
import se.gustavkarlsson.skylight.android.entities.Darkness
import se.gustavkarlsson.skylight.android.entities.GeomagLocation
import se.gustavkarlsson.skylight.android.entities.KpIndex
import se.gustavkarlsson.skylight.android.entities.Weather
import se.gustavkarlsson.skylight.android.services.ChanceEvaluator

class AuroraReportEvaluator(
	private val kpIndexEvaluator: ChanceEvaluator<KpIndex>,
	private val geomagLocationEvaluator: ChanceEvaluator<GeomagLocation>,
	private val weatherEvaluator: ChanceEvaluator<Weather>,
	private val darknessEvaluator: ChanceEvaluator<Darkness>
) : ChanceEvaluator<AuroraReport> {

	override fun evaluate(value: AuroraReport): Chance {
		val activityChance = value.kpIndex.value
			?.let(kpIndexEvaluator::evaluate) ?: Chance.UNKNOWN
		val locationChance = value.geomagLocation.value
			?.let(geomagLocationEvaluator::evaluate) ?: Chance.UNKNOWN
		val weatherChance = value.weather.value
			?.let(weatherEvaluator::evaluate) ?: Chance.UNKNOWN
		val darknessChance = value.darkness.value
			?.let(darknessEvaluator::evaluate) ?: Chance.UNKNOWN

		val chances = listOf(activityChance, locationChance, weatherChance, darknessChance)

		if (chances.any { !it.isKnown }) {
			return Chance.UNKNOWN
		}

		if (chances.any { !it.isPossible }) {
			return Chance.IMPOSSIBLE
		}

		return listOf(weatherChance, darknessChance, activityChance * locationChance).min()!!
	}
}
